### PR TITLE
[7-zip] update to <21.07>

### LINKED ISSUE #21567: [7-zip] update to <21.07>
Library name: **7-zip**

New version number: **21.06**

Other information that may be useful (release notes, etc...)
https://7-zip.org/
https://www.7-zip.org/download.html
https://sourceforge.net/p/sevenzip/discussion/45797/thread/b463f34259/
https://www.7-zip.org/history.txt

This is for the `Port Update`.

`Tool Update Request` can be found in #19568

### PR DESCRIPTION
**Pull request by FrankXie05**
Fix #21567
Fix #21765
[[7-zip] update to <21.07>

Note : no feature need to test

### CODE REVIEW

File: ports/7zip/vcpkg.json [hunk 1]
```diff
@@ -1,8 +1,18 @@
 {
   "name": "7zip",
-  "version-string": "19.00",
-  "port-version": 4,
+  "version-string": "21.07",
   "description": "Library for archiving file with a high compression ratio.",
   "homepage": "https://www.7-zip.org",
-  "supports": "!(linux | osx | arm | uwp)"
+  "license": "BSD-3-Clause",
```
Comment (Osyotr) [inline ports/7zip/vcpkg.json:6]: From `DOC/Licence.txt`:
```
  7-Zip Copyright (C) 1999-2020 Igor Pavlov.

  The licenses for files are:

    1) CPP/7zip/Compress/Rar* files: the "GNU LGPL" with "unRAR license restriction"
    2) CPP/7zip/Compress/LzfseDecoder.cpp: the "BSD 3-clause License"
    3) Some files are "public domain" files, if "public domain" status is stated in source file.
    4) the "GNU LGPL" for all other files. If there is no license information in 
       some source file, that file is under the "GNU LGPL".

  The "GNU LGPL" with "unRAR license restriction" means that you must follow both 
  "GNU LGPL" rules and "unRAR license restriction" rules.
```
Is it possible to express this with `licence` field?

File: ports/7zip/portfile.cmake [hunk 2]
```diff
@@ -14,17 +14,16 @@ vcpkg_extract_source_archive_ex(
...
 
-vcpkg_install_cmake()
+vcpkg_cmake_install()
 
 vcpkg_copy_pdbs()
 
 file(
-    INSTALL ${CMAKE_CURRENT_LIST_DIR}/License.txt
-    DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT}
+    INSTALL "${CMAKE_CURRENT_LIST_DIR}/License.txt"
```
Comment (Osyotr) [inline ports/7zip/portfile.cmake:26]: Licenses for 7zip are in `DOC` folder, no need to keep this file in port folder (also `unRarLicense.txt`).

**Review by Osyotr — CHANGES_REQUESTED** [review]

File: ports/7zip/vcpkg.json (hunk 1 above)
Comment (dg0yt) [reply ports/7zip/vcpkg.json:6]: This is: `"license": null`

File: ports/7zip/portfile.cmake (hunk 2 above)
Comment (Osyotr) [reply ports/7zip/portfile.cmake:26]: Yes, I meant that the actual file `${CMAKE_CURRENT_LIST_DIR}/License.txt` (and `unRarLicense.txt`, which is not used anywhere) should be removed and this line should be replaced with `${SOURCE_PATH}/DOC/License.txt`

**Comment by Osyotr** [comment]
I've noticed that this `CMakeLists.txt` probably does not export targets properly. Could you check that?
Also, there are still leftover files in port directory.

**Comment by FrankXie05** [comment]
@Osyotr Confirm

**Comment by FrankXie05** [comment]
@Osyotr Could you check again？ :)

File: ports/7zip/portfile.cmake [hunk 3]
```diff
@@ -14,17 +14,17 @@ vcpkg_extract_source_archive_ex(
...
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
 )
 
-vcpkg_install_cmake()
+vcpkg_cmake_install()
 
 vcpkg_copy_pdbs()
 
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
```
Comment (Osyotr) [inline ports/7zip/portfile.cmake:25]: vcpkg_cmake_config_fixup?

**Comment by Osyotr** [comment]
Unused license files in port directory are still there.

**Comment by Osyotr** [comment]
Unused license files in port directory are still there.

**Review by LilyWangLL — APPROVED** [review]

**Review by vicroms — APPROVED** [review]

File: ports/7zip/CMakeLists.txt [hunk 4]
```diff
@@ -303,32 +286,114 @@ set(SRC_OTHER
...
-    ${SRC_CRYPTO}
-    ${SRC_C}
     CPP/7zip/Archive/Archive2.def
     C/Util/LzmaLib/LzmaLib.def
 )
+# 7zCrcOpt
+if(VCPKG_TARGET_IS_WINDOWS)
+    if(VCPKG_TARGET_ARCHITECTURE STREQUAL "x86" OR VCPKG_TARGET_ARCHITECTURE STREQUAL "x64")
+        target_sources(7zip PRIVATE Asm/x86/7zCrcOpt.asm)
+        set_source_files_properties(Asm/x86/7zCrcOpt.asm PROPERTIES LANGUAGE ASM_MASM)
```
Comment (vicroms) [inline ports/7zip/CMakeLists.txt:297]: We can reduce the amount of `if` blocks by bundling them together like:

```
target_sources(7zip PRIVATE Asm/x86/7zCrcOpt.asm Asm/x86/AesOpt.asm Asm/x86/Sha1Opt.asm)
set_source_files_properties(Asm/x86/7zCrcOpt.asm Asm/x86/AesOpt.asm Asm/x86/Sha1Opt.asm PROPERTIES LANGUAGE ASM_MASM)
```

**Pull request merged**